query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/QueryRequestToPromqlConverter.java

```diff
@@ -0,0 +1,255 @@
+            .map(
+                functionExpression -> {
+                  String functionName = getFunctionName(functionExpression);
+                  MetricConfig metricConfig = getMetricConfigForFunction(functionExpression);
+                  return buildQuery(
+                      metricConfig.getName(),
+                      functionName,
+                      groupByList,
+                      String.join(", ", filterList),
+                      queryTimeRange.getDuration().toMillis());
+                })
+            .collect(Collectors.toUnmodifiableList());
+
+    return new PromqlQuery(
+        queries,
+        executionContext.getTimeSeriesPeriod().isEmpty(),
+        queryTimeRange.getStartTime(),
+        queryTimeRange.getEndTime(),
+        getStep(executionContext));
+  }
+
+  private String getGroupByList(QueryRequest queryRequest) {
+    List<String> groupByList = Lists.newArrayList();
+    if (queryRequest.getGroupByCount() > 0) {
+      for (Expression expression : queryRequest.getGroupByList()) {
+        // skip datetime function in group-by
+        if (QueryRequestUtil.isDateTimeFunction(expression)) {
+          continue;
+        }
+        groupByList.add(convertColumnAttribute2String(expression));
+      }
+    }
+    return String.join(", ", groupByList);
+  }
+
+  private long getStep(ExecutionContext executionContext) {
+    return executionContext.getTimeSeriesPeriod().map(Duration::toMillis).orElse(-1L);
+  }
+
+  private String buildQuery(
+      String metricName, String function, String groupByList, String filter, long durationMillis) {
+    String template =
+        "%s by (%s) (%s(%s{%s}[%sms]))"; // sum by (a1, a2) (sum_over_time(num_calls{a4="..",
+    // a5=".."}[xms]))
+
+    return String.format(
+        template,
+        function,
+        groupByList,
+        function + "_over_time", // assuming gauge type of metric
+        metricName,
+        filter,
+        durationMillis);
+  }
+
+  private String getFunctionName(Expression functionSelection) {
+    switch (functionSelection.getFunction().getFunctionName().toUpperCase()) {
+      case QUERY_FUNCTION_SUM:
+        return "sum";
+      case QUERY_FUNCTION_MAX:
+        return "max";
+      case QUERY_FUNCTION_MIN:
+        return "min";
+      case QUERY_FUNCTION_AVG:
+        return "avg";
+      case QUERY_FUNCTION_COUNT:
+        return "count";
+      default:
+        throw new RuntimeException("");
+    }
+  }
+
+  private MetricConfig getMetricConfigForFunction(Expression functionSelection) {
+    return prometheusViewDefinition.getMetricConfig(
+        getLogicalColumnName(functionSelection.getFunction().getArgumentsList().get(0)));
+  }
+
+  private String convertColumnAttribute2String(Expression expression) {
+    return prometheusViewDefinition.getPhysicalColumnName(getLogicalColumnName(expression));
+  }
+
+  private String getLogicalColumnName(Expression expression) {
```

Comment:
Is this function present at multiple places ? If yes, can we move it to a common place and import everywhere ?